File: go/4.寻找两个正序数组的中位数.go
package main

/*
 * @lc app=leetcode.cn id=4 lang=golang
 *
 * [4] 寻找两个正序数组的中位数
 */

// @lc code=start
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	var num1point int = 0
	var num2point int = 0
	var mid int = (len(nums1) + len(nums2) + 1) / 2
	evenNumber := ((len(nums1)+len(nums2))&1 == 0)
	var val float64 = 0
	nums1end := false
	nums2end := false
	if len(nums2) == 0 {
		if len(nums1) == 1 {
			return float64(nums1[0])
		}
		if evenNumber {
			return float64(nums1[mid]+nums1[mid-1]) / 2
		} else {
			return float64(nums1[mid-1])
		}

	}
	if len(nums1) == 0 {
		if len(nums2) == 1 {
			return float64(nums2[0])
		}
		if evenNumber {
			return float64(nums2[mid]+nums2[mid-1]) / 2
		} else {
			return float64(nums2[mid-1])
		}
	}

	var lastVal int = 0

	//println("mid: ", mid, "evenNum: ", evenNumber)
	for !nums1end || !nums2end {
		additional := 0
		if nums1end {
			additional += 1
		}
		if nums2end {
			additional += 1
		}
		//println("point1: ", num1point, "point2: ", num2point, "additional: ", additional, "lastVal: ", lastVal, "val: ", val)
		if num1point+num2point+additional == mid {
			if !evenNumber {
				//println("11")
				return float64(lastVal)
			}
			val = float64(lastVal)
		}

		if num1point+num2point+additional == mid+1 {
			val = (val + float64(lastVal)) / 2
			// println("22")

			return val
		}
		//println(nums1end, nums2end)
		if (nums1[num1point] <= nums2[num2point] && !nums1end) || nums2end {
			lastVal = nums1[num1point]

			if num1point == len(nums1)-1 {
				nums1end = true
			} else {
				num1point++
			}
		} else {
			lastVal = nums2[num2point]

			if num2point == len(nums2)-1 {
				nums2end = true
			} else {
				num2point++
			}
		}
	}
	//println("33")

	return (val + float64(lastVal)) / 2
}

// @lc code=end

File: c/4.寻找两个正序数组的中位数.c
/*
 * @lc app=leetcode.cn id=4 lang=c
 *
 * [4] 寻找两个正序数组的中位数
 */

// @lc code=start
#include<stdlib.h>
#include<stdio.h>

double array_value(int* nums1, int nums1Size, int* nums2, int index){
    if(index < nums1Size){
        return (double)nums1[index];
    }
    return (double)nums2[index - nums1Size];

}


double findMedianSortedArrays(int* nums1, int nums1Size, int* nums2, int nums2Size){
    int *index_num = (int*) calloc(nums1Size + nums2Size, sizeof(int));
    if(nums1Size == 0 && nums2Size == 0){
        return 0.0;
    }
    if(nums1Size == 0 && nums2Size == 1){
        return (double)nums2[0];
    }

    if(nums1Size == 1 && nums2Size == 0){
        return (double)nums1[0];
    }

    if(nums1Size == 1 && nums2Size == 1){
        return ((double)nums1[0] + (double)nums2[0]) / 2;
    }


    int median = (nums1Size + nums2Size) / 2;
    int median_second = 0;
    int has_second = 0;
    if ((nums1Size + nums2Size) % 2 == 0)
    {
        has_second = 1;
        median_second = median - 1;
    }

    int num1_index = nums1Size / 2;
    int num2_index = nums2Size / 2;

    while (1)
    {
        if(index_num[median] != 0){
            if(has_second){
                if(index_num[median_second] != 0){
                    return (array_value(nums1, nums1Size, nums2, index_num[median]) + array_value(nums1, nums1Size, nums2, index_num[median - 1])) / 2;
                }
            }else{
                return array_value(nums1, nums1Size, nums2, index_num[median]);
            }
        }

        if(nums1[num1_index] < nums2[num2_index]){
            int tmp_index = num1_index + num2_index;
            index_num[tmp_index] = num1_index;
            //printf("index: %d is num1_index: %d, value is: %d\n", tmp_index, num1_index, nums1[num1_index]);
            if(tmp_index > median)
            {
                num2_index -= 1;
            }
            if(tmp_index < median)
            {
                num1_index += 1;
            }
            if(tmp_index == median && has_second){
                num2_index -= 1;
            }
        }else if (nums1[num1_index] > nums2[num2_index])
        {
            int tmp_index = num1_index + num2_index;
            index_num[tmp_index] = nums1Size + num2_index;
            //printf("index: %d is num2_index: %d, value is: %d\n", tmp_index, (num2_index + nums1Size), nums2[num2_index]);
            if(tmp_index > median)
            {
                num1_index -= 1;
            }
            if(tmp_index < median)
            {
                num2_index += 1;
            }
            if(tmp_index == median && has_second){
                num1_index -= 1;
            }
        }
        
    }
    



}
// @lc code=end

int main()
{
    int nums1[] = {1,2};
    int nums2[] = {3,4};
    int nums1Size = 2;
    int nums2Size = 2;
    double result = findMedianSortedArrays(nums1, nums1Size, nums2, nums2Size);
    printf("%lf", result);
    return 0;
}